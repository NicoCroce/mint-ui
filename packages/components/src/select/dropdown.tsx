import * as React from "react";
import styled, { css } from "styled-components";
import COLORS from "../__utils/colors";
import { GlobalStyles } from "../app";

export const DropdownWrapper = styled.div`
	${GlobalStyles};
	margin: 0;
	padding: 0;
	list-style: none;
	background-color: ${COLORS.WHITE};
	border-radius: 4px;
	outline: none;
	-webkit-box-shadow: 0 2px 8px rgba(0, 0, 0, 0.15);
	box-shadow: 0 2px 8px rgba(0, 0, 0, 0.15);
`;

export const List = styled.ul`
	border: 1px solid ${COLORS.BACKGROUND_GREY};
	border-radius: 4px;
	list-style: none;
	margin: 0;
	padding: 0;
	max-height: 250px;
	margin-bottom: 0;
	padding-left: 0;
	overflow: auto;
	outline: none;
`;

export interface OptionProps {
	key?: string | number;
	disabled?: boolean;
	value?: string | number;
	title?: string;
	children?: React.ReactNode;
	className?: string;
	style?: React.CSSProperties;
	onClick?: (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => void;
}

export const StyledOption = styled.li<Pick<OptionProps, "disabled">>`
	display: block;
	padding: 5px 12px;
	overflow: hidden;
	color: ${COLORS.GREY1};
	font-weight: normal;
	line-height: 22px;
	white-space: nowrap;
	text-overflow: ellipsis;
	cursor: pointer;
	-webkit-transition: background 0.3s ease;
	transition: background 0.3s ease;
	${props =>
		props.disabled &&
		css`
			cursor: no-drop;
		`}
	&:first-child {
		color: rgba(0, 0, 0, 0.65);
		font-weight: 600;
		background-color: #fafafa;
	}
	&:hover {
		background-color: #e6f7ff;
	}
`;

export interface SelectedOption {
	value: string | number;
	title: string;
}
export const Option = ({
	key,
	value,
	disabled = false,
	children,
	onClick
}: OptionProps) => {
	return (
		<StyledOption
			key={key}
			value={value}
			disabled={disabled}
			onClick={onClick}
		>
			{children}
		</StyledOption>
	);
};

interface DropdownProps {
	onSelect?: (option: SelectedOption) => void;
	children:
	| React.ComponentElement<any, any>
	| React.ComponentElement<any, any>[];
}

export const Dropdown = ({ onSelect, children }: DropdownProps) => {
	const handleSelect = (event: React.ChangeEvent<HTMLDivElement>) => {
		const title = event.target.innerText || event.target.textContent || "";
		const value = event.target.getAttribute("value") || "";
		if (event.target) {
			onSelect ? onSelect({ value, title }) : null;
		}
	};
	return (
		<DropdownWrapper data-testid="dropdown">
			<List>
				{React.Children.map(
					children,
					(child: React.CElement<any, any>) =>
						React.cloneElement(child, {
							onClick: handleSelect
						})
				)}
			</List>
		</DropdownWrapper>
	);
};

export default Dropdown;
