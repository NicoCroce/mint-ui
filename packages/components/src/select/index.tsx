import * as React from "react";
import styled from "styled-components";
import { Colors } from "@turtlemint/core";
import ExpandMore from "../icons/expand-more";
import ExpandLess from "../icons/expand-less";
import Dropdown, { SelectedOption } from "./dropdown";
import SIZE from "../__utils/size";

export const SelectWrapper = styled.div`
	width: 300px;
	box-sizing: border-box;
`;
export const StyledSelectCTA = styled.div`
	box-sizing: border-box;
	outline: none;
	border: 0;
	padding: 8px 16px;
	border: 1px solid ${Colors.DISABLED};
	border-radius: 4px;
	font-size: ${SIZE.INPUT_FONT_SIZE};
	color: ${Colors.GREY1};
	max-width: 300px;
	&:hover {
		border: 1px solid ${Colors.PERSIAN_GREEN};
		cursor: pointer;
	}
	display: flex;
	justify-content: space-between;
	align-items: center;
`;

const ArrowWrapper = styled.span`
	color: ${Colors.GREY1};
`;

const ArrowToggle = ({ open }: { open: boolean }) => (
	<ArrowWrapper>{!open ? <ExpandMore /> : <ExpandLess />}</ArrowWrapper>
);

interface SelectCTAProps {
	value: string;
	showArrow?: boolean;
	open?: boolean;
	onClick?: () => void;
}

export const SelectCTA = ({
	value,
	showArrow = false,
	open,
	onClick
}: SelectCTAProps) => {
	const handleSelectCTAClick = () => {
		onClick ? onClick() : null;
	};
	return (
		<StyledSelectCTA onClick={handleSelectCTAClick}>
			<span>{value}</span>
			{showArrow ? <ArrowToggle open={open as boolean} /> : null}
		</StyledSelectCTA>
	);
};

interface SelectProps {
	value: string;
	onSelect?: (option: SelectedOption) => void;
	children:
		| React.ComponentElement<any, any>
		| React.ComponentElement<any, any>[];
}

export const Select = ({ value, onSelect, children }: SelectProps) => {
	const [open, setOpen] = React.useState(false);
	const handleSelect = (option: SelectedOption) => {
		setOpen(!open);
		onSelect ? onSelect(option) : null;
	};
	return (
		<SelectWrapper data-testid="select-wrapper">
			<SelectCTA
				open={open}
				value={value}
				showArrow={true}
				onClick={() => setOpen(!open)}
			/>
			{open ? (
				<Dropdown onSelect={handleSelect}>{children}</Dropdown>
			) : null}
		</SelectWrapper>
	);
};

export default Select;
