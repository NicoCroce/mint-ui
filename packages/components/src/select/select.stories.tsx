import * as React from "react";
import { storiesOf } from "@storybook/react";
import Select from "./";
import { Option } from "./dropdown";

const stories = storiesOf("Select", module);

export const data = [
    {
        value: 1,
        text: "Item 1"
    },
    {
        value: 2,
        text: "Item 2"
    }
];
interface DemoSelectProps {
    onSelect?: (value: string) => void;
}
export const DemoSelect = ({ onSelect }: DemoSelectProps) => {
    const [value, setValue] = React.useState("Select Value");

    const handleSelect = (option: any) => {
        setValue(option.title);
        //for test case
        onSelect(option.title);
    }
    return (
        <Select value={value} onSelect={handleSelect}>
            {
                data.map((d: any) => (
                    <Option
                        key={d.value}
                        value={d.value}>
                        {d.text}
                    </Option>
                ))
            }
        </Select>
    );
}
stories.add("full functional demo", () => (
    <DemoSelect />
))

stories.add("default", () => (
    <Select value="Test value">
        {
            data.map((d: any) => (
                <Option
                    key={d.value}
                    value={d.value}>
                    {d.text}
                </Option>
            ))
        }
    </Select>
))