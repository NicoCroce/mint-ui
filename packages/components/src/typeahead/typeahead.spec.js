import * as React from "react";
import { render } from "@testing-library/react";
// import axiosMock from "axios";
import TypeAhead from "./index";
import { Option } from "../select/dropdown";

export const data = [
	{
		value: 1,
		text: "Item 1"
	},
	{
		value: 2,
		text: "Item 2"
	}
];

// export const TestTypeAhead = () => {
//     const [data, setData] = React.useState([]);
//     const [open, setOpen] = React.useState(false);
//     const [fetching, setFetching] = React.useState(false);

//     const fetchUser = async () => {
//         setFetching(true);
//         setOpen(false);
//         const response = await axios.get('https://randomuser.me/api/?results=15');
//         const data = response.data.results.map((user: any) => ({
//             text: `${user.name.first} ${user.name.last}`,
//             value: user.login.username,
//         }));
//         setData(data);
//         setOpen(true);
//         setFetching(false);
//     }

//     const handleSelect = (value: unknown) => {
//         console.log(value);
//         setOpen(false);
//     }

//     return (
//         <Row style={{
//             padding: "0 30px"
//         }}>
//             <Col>
//                 <TypeAhead
//                     value="some value"
//                     loading={fetching}
//                     fetchFunc={fetchUser}
//                     onSelect={handleSelect}
//                     open={open}
//                     placeholder="Select user..."
//                 >
//                     {data.map((d: any) => (
//                         <Option key={d.value} value={d.value}>{d.text}</Option>
//                     ))}
//                 </TypeAhead>
//             </Col>
//         </Row>
//     )
// }
describe("Typeahead tests", () => {
	test("renders correctly", () => {
		const { asFragment } = render(
			<TypeAhead
				value="some value"
				loading={false}
				open={true}
				placeholder="Select user..."
			>
				{data.map(d => (
					<Option key={d.value} value={d.value}>
						{d.text}
					</Option>
				))}
			</TypeAhead>
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
