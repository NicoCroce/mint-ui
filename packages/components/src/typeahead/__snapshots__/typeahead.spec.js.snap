// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typeahead tests renders correctly 1`] = `
<DocumentFragment>
  <div
    class="select__SelectWrapper-sc-1trodvu-0 jtVoAg"
    tabindex="0"
  >
    <div
      class="select__StyledSelectCTA-sc-1trodvu-1 bFMGxn"
    >
      <span>
        some value
      </span>
    </div>
    <div
      class="dropdown__DropdownWrapper-sc-4mjgs1-0 jNYjVA"
      data-testid="dropdown"
    >
      <ul
        class="dropdown__List-sc-4mjgs1-1 dWFdKM"
      >
        <li
          class="dropdown__StyledOption-sc-4mjgs1-2 iFCPBi"
          value="1"
        >
          Item 1
        </li>
        <li
          class="dropdown__StyledOption-sc-4mjgs1-2 iFCPBi"
          value="2"
        >
          Item 2
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
